In [11]:
# para que los cambios en .scr se reflejen en el notebook
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [12]:
# para usar los modulos de src
import sys
from pathlib import Path
notebook_path = Path().resolve()
project_root = notebook_path.parent
sys.path.append(str(project_root))

# carga las variables de entorno
from dotenv import load_dotenv
load_dotenv()

False

# EDA

## librerias necesarias

In [36]:
from src.utils.project_paths import RAW_DATA_DIR, ROOT_DIR, FIGURES_DIR
from src.utils.plot_styles import PaletaPlotly
import pandas as pd
# para graficos
import plotly
import kaleido
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import plotly.io as pio 
import seaborn as sns

In [14]:
paleta = PaletaPlotly()

### cargar los archivos

In [15]:
data_file_path = RAW_DATA_DIR / 'CARTERA_PRUEBA_TECNICA.csv'
if data_file_path.exists():
    df_cartera = pd.read_csv(data_file_path)
else:
    print(f'No existe el archivo {data_file_path}')

In [16]:
# vistazo a los datos
print(df_cartera.shape)
df_cartera.head()

(21499, 41)


aniomes    telefono  account_num  target  vnc_no  vencimiento  \
0   202401  1773210294   1812120285       0     0.0          0.0   
1   202312  1284496476   1258714049       0     1.0          0.0   
2   202311  1312772779   1745686591       1     1.0          1.0   
3   202312  1886941825   1930858772       1     1.0          1.0   
4   202312  1124231538   1411143386       0     1.0          0.0   

   decil_max_ultimos3m_totalmonto  decil_max_ultimos3m_pago      forma_pago  \
0                               9                         9     DOMICILIADO   
1                               2                         2     DOMICILIADO   
2                               4                         4  NO_DOMICILIADO   
3                               2                         2  NO_DOMICILIADO   
4                               7                         7  NO_DOMICILIADO   

   datos_curso_q  ... max_trimestral_sobrefacturacion  pct_datos_4g  \
0              1  ...                        0.871757      0.966336   
1              2  ...                        1.044287      0.000000   
2              3  ...                        0.851603      0.967840   
3              3  ...                        0.892761      0.476536   
4              4  ...                        0.892946      0.923823   

   no_exitoso  monto_total_cp_ultimomes  promedio_semestral_sobrefacturacion  \
0        87.0                     36.69                             0.826413   
1        27.0                     10.99                             0.000000   
2       100.0                     13.03                             0.833982   
3       120.0                     11.19                             0.892761   
4        25.0                     22.23                             0.885270   

   cantidad_minutos_mes  max_ultimos3m_pago  n_entrantes_offnet  \
0            513.449994              113.69                  63   
1              3.866667               33.87                   6   
2             33.233334               40.23                  29   
3            184.533333               33.57                 153   
4              5.733333               62.21                   5   

   monto_pago_cp_ultimomes  conteo_vecesen_finterm  
0                    36.69                       0  
1                    10.99                       0  
2                    13.03                       0  
3                    11.19                       0  
4                    22.23                       0  

[5 rows x 41 columns]

In [17]:
df_cartera.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21499 entries, 0 to 21498
Data columns (total 41 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   aniomes                              21499 non-null  int64  
 1   telefono                             21499 non-null  int64  
 2   account_num                          21499 non-null  int64  
 3   target                               21499 non-null  int64  
 4   vnc_no                               20424 non-null  float64
 5   vencimiento                          20424 non-null  float64
 6   decil_max_ultimos3m_totalmonto       21499 non-null  int64  
 7   decil_max_ultimos3m_pago             21499 non-null  int64  
 8   forma_pago                           21499 non-null  object 
 9   datos_curso_q                        21499 non-null  int64  
 10  susp_cobranza                        21499 non-null  object 
 11  max_diferencia_dias_ultimos3

In [35]:
df_cartera.drop(columns=['aniomes','telefono','account_num','target']).describe()

vnc_no   vencimiento  decil_max_ultimos3m_totalmonto  \
count  20424.000000  20424.000000                    21499.000000   
mean       0.884694      0.625685                        5.891111   
std        0.319398      0.483957                        3.669630   
min        0.000000      0.000000                        1.000000   
25%        1.000000      0.000000                        3.000000   
50%        1.000000      1.000000                        6.000000   
75%        1.000000      1.000000                        8.000000   
max        1.000000      1.000000                       20.000000   

       decil_max_ultimos3m_pago  datos_curso_q  max_diferencia_dias_ultimos3m  \
count              21499.000000   21499.000000                   21499.000000   
mean                   5.885250       2.785060                      25.632215   
std                    3.659754       1.350683                      28.978778   
min                    1.000000       1.000000                    -310.000000   
25%                    3.000000       2.000000                      13.000000   
50%                    6.000000       3.000000                      17.000000   
75%                    8.000000       4.000000                      28.000000   
max                   20.000000       5.000000                     649.000000   

         n_ses_datos  promedio_ultimos3m_finterm  pct_datos_3g    datos_curso  \
count   21499.000000                21499.000000  21499.000000   21499.000000   
mean     4095.824131                   14.000217      0.146385   10286.601360   
std      3474.424103                   84.588692      0.252551   12437.778407   
min         0.000000                   -0.000460      0.000000       0.000000   
25%      1969.500000                    0.000000      0.013229    3149.491606   
50%      3517.000000                    0.000000      0.046908    7549.402596   
75%      5425.000000                    0.000000      0.133215   13574.340593   
max    153347.000000                 2649.000000      1.000000  493711.993183   

       ...  max_trimestral_sobrefacturacion  pct_datos_4g    no_exitoso  \
count  ...                     21499.000000  21499.000000  21499.000000   
mean   ...                         0.882989      0.813277    102.479070   
std    ...                         6.157047      0.300584    127.309622   
min    ...                        -0.131485      0.000000      0.000000   
25%    ...                         0.702985      0.833105     35.000000   
50%    ...                         0.889295      0.944180     74.000000   
75%    ...                         0.898774      0.982295    130.000000   
max    ...                       893.000000      1.000000   6209.000000   

       monto_total_cp_ultimomes  promedio_semestral_sobrefacturacion  \
count              21499.000000                         21499.000000   
mean                  25.520961                             0.749510   
std                   53.388678                             0.524997   
min                    0.000000                            -1.721921   
25%                   12.790000                             0.410532   
50%                   16.800000                             0.862421   
75%                   25.570000                             0.893220   
max                 2665.600000                            25.273585   

       cantidad_minutos_mes  max_ultimos3m_pago  n_entrantes_offnet  \
count          21499.000000        21499.000000        21499.000000   
mean             172.016371           66.810356           53.939020   
std              310.362505           75.396365           64.980932   
min                0.000000           -0.001270            0.000000   
25%               31.491667           35.810000           18.000000   
50%               88.866667           49.480000           37.000000   
75%              201.858333           73.095000           70.000000   
max            11877.099931         26

El dataset contiene 40 dimensiones y 21499 filas con lo cual se tiene una buena proporcion de datos para realizar un modelo de machinne learnign
ademas el dataset no contiene datos nulos
se debe realizar transformaciones de datos para los atributos que son de tipo string pero a pesar de ello no habria mayor problema

## Distribucion de los clientes que cayeron en mora

In [18]:
series_target_count=df_cartera.value_counts('target')
series_target_percentage=df_cartera.value_counts('target',normalize=True).mul(100).round(2)
df_target=pd.concat([series_target_count,series_target_percentage],axis=1).reset_index()
df_target['target']=df_target['target'].map({0:'No',1:'Si'})
df_target

target  count  proportion
0     Si  11415        53.1
1     No  10084        46.9

In [101]:
fig = make_subplots(rows=1, cols=2, subplot_titles=('Cantidad', 'Porcentual'))

fig1 = px.bar(df_target, x='target', y='count', title='Target Count',color_discrete_sequence=[paleta.color_blue])
fig2 = px.bar(df_target, x='target', y='proportion', title='Target Proportion',color_discrete_sequence=[paleta.color_blue])
fig.add_trace(fig1.data[0], row=1, col=1)
fig.add_trace(fig2.data[0], row=1, col=2)

fig.update_traces(texttemplate='%{y}', textposition='outside',textfont_size=12,row=1, col=1)
fig.update_traces(texttemplate='%{y:.1f}%', textposition='outside', textfont_size=12, row=1, col=2)

fig.update_yaxes(title_text='Clientes (Q)', row=1, col=1)
fig.update_yaxes(title_text='Clientes (%)', row=1, col=2)

fig.update_layout(
    height=600, 
    width=1000,
    title_text="Distribución de Clientes que caen en Cartera", 
    title_x=0.5,  
    title_font_size=16  
)
fig.show()
fig.write_image(str(FIGURES_DIR / 'clientes_mora_distribution.svg'))
fig.write_image(str(FIGURES_DIR / 'clientes_mora_distribution.png'))

- existe un 3% mayor de datos de los clientes que cayeron en mora
- para un modelo de machine learnign no habria mayor problema en el desbalance de datos

In [141]:
series_mora_time_percetage=df_cartera.groupby(['aniomes'])['target'].value_counts(normalize=True).mul(100).round(2)
series_mora_time_count=df_cartera.groupby(['aniomes'])['target'].value_counts()
df_mora_time=pd.concat([series_mora_time_count,series_mora_time_percetage],axis=1).reset_index()
df_mora_time['target']=df_mora_time['target'].map({0:'No',1:'Si'})
df_mora_time['aniomes'] = df_mora_time['aniomes'].astype(str)
df_mora_time

aniomes target  count  proportion
0  202311     Si   4432       64.43
1  202311     No   2447       35.57
2  202312     No   3699       53.34
3  202312     Si   3236       46.66
4  202401     No   3938       51.24
5  202401     Si   3747       48.76

In [143]:
# df_mora_time['fecha'] = pd.to_datetime(df_mora_time['aniomes'].astype(str), format='%Y%m')
fig = px.line(df_mora_time.sort_values('aniomes'), x='aniomes', y='proportion', color='target',
              title='Target Proportion', 
              text='count',
              color_discrete_sequence=[paleta.color_red, paleta.color_grey_3])

# Añadir marcadores (puntos) en las líneas
fig.update_traces(mode='lines+markers+text', textposition='top center', textfont_size=12)

# Formatear el texto para mostrar porcentaje y count
fig.update_traces(texttemplate='%{y:.1f} (%{text}%)')

# Actualizar la leyenda
fig.update_layout(legend_title_text='En Mora')

# Actualizar etiquetas de los ejes
fig.update_yaxes(title_text='Clientes (%)')
fig.update_xaxes(title_text='Periodo')

# Actualizar el diseño general
fig.update_layout(
    height=500, 
    width=1000,
    title_text="Distribución de Clientes que caen en Cartera por Mes", 
    title_x=0.5,  
    title_font_size=16  
)
# ajustar el eje y
fig.update_yaxes(range=[30, 70])
# fig.update_xaxes(range=['2023-10-21', '2024-01-15'])

fig.show()

In [144]:
fig = px.bar(df_mora_time, x='aniomes', y='proportion', title='Target Proportion',color='target',barmode='group',text='count',color_discrete_sequence=[paleta.color_red,paleta.color_grey_3])
fig.update_traces(texttemplate='%{y:.1f}%<br>(%{text})', textposition='outside', textfont_size=12)
# legend title
fig.update_layout(legend_title_text='En Mora')
fig.update_yaxes(title_text='Clientes (%)')
fig.update_xaxes(title_text='Periodo')
fig.update_layout(
    height=800, 
    width=1000,
    title_text="Distribución de Clientes que caen en Cartera por Mes", 
    title_x=0.5,  
    title_font_size=16  
)
fig.show()
fig.write_image(str(FIGURES_DIR / 'clientes_mora_distribution_time.svg'))
fig.write_image(str(FIGURES_DIR / 'clientes_mora_distribution_time.png'))


## Arquetipo de la Persona que ha caido en mora

- graficar las salidas por mes
- graficar el target, en cantidad y proporcion
- histograma de la edad (target)
- histograma de arpu (target)
- histograma de la edad(target)
